### ISSUE TITLE
Orders Sync'ing: Delete remotely unavailable entities

### ISSUE THREAD

**Issue by jleandroperez** [woocommerce/woocommerce-ios#291]
### Details:
As of now, Orders Sync'ing is always additive. Once an entity is locally cached, we never actually nuke it.

We need to implement logic to detect the following scenario. One (UNTESTED / UNVERIFIED) alternative that comes to mind is the following:

A. "Orders Page" of size N
B. Cached order M
C. Should M be present in the "Orders Page"? (could probably be determined by means of the Order ID)

If (C) is true, and yet, the order is missing, we should proceed to deleting the local copy.

cc @bummytime @mindgraffiti  any thoughts would be more than welcome!!

**Comment by mindgraffiti** [comment]
@rachelmcr thanks for posting the feedback! Per p99K0U-1Fs-p2 the decision was made to fix this problem (and several others relating to syncing) after I'm done with the Refunds feature. My current best guess is that a fix is 1-2 months out.

**Comment by mindgraffiti** [comment]
2409923-zen is experiencing this issue as well. @melindahelt has already responded.

**Comment by mindgraffiti** [comment]
2413570-zen reported this issue as well

**Comment by mindgraffiti** [comment]
Refunds has been a ... challenging feature. I will raise this issue again and see if it can be prioritized for pick up by another iOS dev.

**Comment by joshbetz** [comment]
@woocommerce/ios-developers Does anyone know if this still happens? I tried on iOS and couldn't reproduce it.

**Comment by shiki** [comment]
@joshbetz sorry for the delay. This is still possible but it's probably less likely to happen now. In https://github.com/woocommerce/woocommerce-ios/pull/2017, I added **deletion of all orders** during synchronization. That resets everything including the stale data. 

This is a hack. And it degrades the performance because, in addition to deleting all the orders, we're also performing dual fetches. Rationales are explained in the PR: https://github.com/woocommerce/woocommerce-ios/pull/2017. And the code also has this documentation: [`OrderListSyncActionUseCase.swift`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Orders/OrderListSyncActionUseCase.swift). 

The same problem should exist with Products. We also delete all the products when pulling to refresh. 

Android has a better sync strategy: p99K0U-28U-p2. But it's quite complex and there are probably better ways to do it for iOS. We did consider this during the Slow Orders project (p91TBi-4p5-p2) but have enough time. 

There were some discussions here about how we can improve the sync strategy for all areas: p7bje6-1IQ-p2. There were also some discussions about supporting tombstoning in the API but I can't find it. Maybe it's buried in the comments of that post. 😅 

In summary, I think we can close this issue. And we can revisit how we can do better syncing at some point. What do you think?

**Comment by joshbetz** [comment]
Ok, this is super helpful context. I did see some of the deletion of all orders logic you mentioned and wondered why we do that 😅

I'll defer to you about whether to keep this open. My thinking would be to keep it open in the backlog for now, but don't feel super strongly about it.

**Comment by shiki** [comment]
Okay! We'll keep it open. 🙂

**Comment by joshbetz** [comment]
What if instead of doing a full sync, we only look at orders that have been modified since the last sync, using the new `modified_after` parameter that's available in WooCommerce 5.8? Moving an order to the trash also updates the `post_modified_gmt` timestamp, so should be included in the API response.

In WooCommerce versions prior to 5.8, we could query all trashed posts (`?status=trash`) and remove them accordingly.

We also have to account for permanently deleted orders. I know previously there has been discussion about tombstones, but I wonder if we can do something simpler. Instead of recording when each individual order is deleted, we could record the timestamp of the last time ANY order was permanently deleted by a person. Then we could check that value in the app (via some API endpoint) and do a full sync. It would act sort of as a cache flush. We wouldn't count orders that are automatically permanently deleted after 30 days because that should be enough time for the deletion to get recorded in the app. If it has ever been more than 30 days since the last sync, we probably want to do a full sync at that point anyway?

The assumption I'm making is that manually permanently deleting orders is somewhat rare since this happens in the background automatically. The benefit of this over tombstones is that it's much easier to implement -- we just have to record a timestamp in an option and build a way for the app to get that data. No new tables would be required. The downside is that if someone is constantly going in and permanently deleting orders, the benefits are more limited (but still better than what we have now).

The implementation in the app should also be fairly straightforward. When we sync, we only update modified posts instead of all posts. If the cache was flushed (a manual permanent delete) since the last time we synced, we purge the synced posts on device and do a full sync.

**Comment by joshbetz** [comment]
Or we put tombstones in an option. We could keep an array of permanently deleted IDs.

**Comment by shiki** [comment]
Related: p1634846145011400-slack-CGPNUU63E.

**Comment by JorgeMucientes** [comment]
Closing this as part of a maintenance pass on the GitHub issue backlog. This issue matched the auto-close criteria for this pass: it has not been updated since 2021-10-22, is unassigned, is not labeled as a bug or high/medium priority, and does not appear to involve POS, payments, login, data loss, crash, security, vulnerability, auth, or fraud.

If this is still relevant, please comment with current details and we can reopen it or move it into Linear for triage.
